Post data to Kafka Events from created/updated in DB

imcaizheng opened the pull request:
For verification, I tried with the following requests in Postman:

```
create job with booking manager
get job with booking manager
search jobs with booking manager
put job with booking manager
patch job with booking manager

create job candidate with booking manager
get job candidate with booking manager
search job candidates with booking manager
put job candidate with booking manager
patch job candidate with booking manager

create resource booking with booking manager
get resource booking with booking manager
search resource bookings with booking manager
put resource booking with booking manager
patch resource booking with booking manager

GET /taas-teams with booking manager
GET /taas-teams/:id with booking manager
```

The only difference is the responses from requests `get job with booking manager` and `get job candidate with booking manager`
would include new fields `updatedAt` and `updatedBy` with null value:


1. Here is the response from the request `get job with booking manager`:

    ``` json
    {
        "id": "45d4d3fb-b7b2-45df-bd31-190bc5848577",
        "projectId": 111,
        "externalId": "1212",
        "description": "Dummy Description",
        "startDate": "2020-09-27T04:17:23.131Z",
        "endDate": "2020-09-27T04:17:23.131Z",
        "numPositions": 13,
        "resourceType": "Dummy Resource Type",
        "rateType": "hourly",
        "workload": "full-time",
        "skills": [
            "23e00d92-207a-4b5b-b3c9-4c5662644941",
            "7d076384-ccf6-4e43-a45d-1b24b1e624aa",
            "cbac57a3-7180-4316-8769-73af64893158",
            "a2b4bc11-c641-4a19-9eb7-33980378f82e"
        ],
        "title": "Dummy title - at most 64 characters",
        "status": "sourcing",
        "createdAt": "2021-02-06T09:10:57.971Z",
        "createdBy": "57646ff9-1cd3-4d3c-88ba-eb09a395366c",
        "updatedAt": null,
        "updatedBy": null
    }
    ```

2. Here is the response from the request `get job candidate with booking manager`:

    ``` json
    {
        "id": "aa03c0fd-0e8e-4e44-ba57-5c7d6bde43d7",
        "jobId": "aa0d81d4-f992-4956-9744-14ec84ecd523",
        "userId": "a55fe1bc-1754-45fa-9adc-cf3d6d7c377a",
        "externalId": "300234321",
        "resume": "http://example.com",
        "status": "open",
        "createdAt": "2021-02-06T09:05:55.862Z",
        "createdBy": "57646ff9-1cd3-4d3c-88ba-eb09a395366c",
        "updatedAt": null,
        "updatedBy": null
    }
    ```

3. Here is the response from the request `get resource booking with booking manager`. It does not include null values due to the `clearObject` function.

    ``` json
    {
        "id": "753634fe-6064-4fe8-a813-7e89cb2af4a7",
        "projectId": 111,
        "userId": "a55fe1bc-1754-45fa-9adc-cf3d6d7c377a",
        "jobId": "aa0d81d4-f992-4956-9744-14ec84ecd523",
        "startDate": "2020-09-27T04:17:23.131Z",
        "endDate": "2020-09-27T04:17:23.131Z",
        "memberRate": 13.23,
        "customerRate": 13,
        "rateType": "hourly",
        "status": "assigned",
        "createdAt": "2021-02-06T09:06:15.282Z",
        "createdBy": "57646ff9-1cd3-4d3c-88ba-eb09a395366c",
        "updatedAt": "2021-02-06T09:06:23.069Z",
        "updatedBy": "57646ff9-1cd3-4d3c-88ba-eb09a395366c"
    }
    ```

maxceem commented on src/models/index.js:
```diff
@@ -17,11 +17,21 @@ const sequelize = new Sequelize(config.get('DATABASE_URL'), {
 
+// add customized toJSON method to a model
+const addMethodOfToJSON = (model) => {
+  model.prototype.toJSON = function () {
```
@imcaizheng I have some concerns regarding the naming of this method. Sequalize model already has method `toJSON` so this would override it which could potentially lead to confusion. Can we use a different name for this? Perhaps something like `getCleared` or `getCleanObject` to indicate that this method return some cleaned/prepared data of the document.

maxceem replied:
@imcaizheng before making any fix to this there is one idea. Do I understand it right, that if instead of manual control of `deletedAt` we are utilizing built-in `paranoid` mode of sequalize, then `deletedAt` would NOT be returned automatically?

i. e. if we implement https://github.com/topcoder-platform/taas-apis/issues/132, then we don't have to implement a custom method `getCleared` and `deletedAt` would not be returned by the standard `toJSON` or `get`?

imcaizheng replied:
@maxceem You are indeed right. After we implement #132, we don't need to manually remove the `deletedAt` field. Currently this `toJSON` function can be served as a temporary fix.

maxceem replied:
@imcaizheng ok, great. 

Let's implement this together with  #132 so we can test how both fixes work together. You may implement #132 in the same branch as this one `imcaizheng:post-event-with-payload-from-db-data` so all the changes would be included into this PR.

Just in case, there are some merge conflicts after merging another PR.

imcaizheng commented:
- updated index and export scripts(auto exclude deleted records and auto exclude the deleted field)
- updated event handlers(auto exclude deleted records)
- updated search/get/create/update functions (auto exclude deleted records and auto exclude the deleted field)
- updated delete functions to use `instance.destroy()`

imcaizheng commented on src/models/Job.js:
```diff
@@ -108,34 +98,35 @@ module.exports = (sequelize) => {
+      deletedAt: 'deletedAt',
+      createdAt: 'createdAt',
+      updatedAt: 'updatedAt',
+      timestamps: true,
```
timestamps must be enable in favor of paranoid mode. It also means `updatedAt` will be automatically set when creating a record.

maxceem replied:
@imcaizheng I thought this would not be required as documentation didn't mention this. But if we need to enable it this is fine. Actually, I was considering if we have to switch to utilizing `timestamps` options instead of manual settings these values but didn't find strong reasons.

The only thing, as we enable `timestamps` we should remove all the code which sets `createdAt` / `updatedAt` dates manually, to keep things clear when we update https://github.com/topcoder-platform/taas-apis/blob/dev/src/services/JobService.js#L202 or create some data https://github.com/topcoder-platform/taas-apis/blob/dev/src/services/JobService.js#L153.

imcaizheng replied:
@maxceem All the redundant code is removed now.

maxceem reviewed (changes requested):
Thank you, @imcaizheng.

The code looks good. But when I'm trying to create a job I'm getting error:
```js
{
    "message": "Internal server error"
}
```
and in console I can see
```
JobService createJob ERROR : column "createdAt" of relation "jobs" does not exist
```

I guess field definitions shouldn't be removed even when using `timestamp`:

![image](https://user-images.githubusercontent.com/146016/107199517-357ea000-69ff-11eb-84c0-a34fb89cbf7b.png)

imcaizheng commented:
@maxceem Fixed.

maxceem reviewed (approved):
Works perfectly now @imcaizheng
